common/src/main/scala/org/apache/celeborn/common/rpc/netty/Inbox.scala

```diff
@@ -59,109 +62,171 @@ private[celeborn] case class RemoteProcessConnectionError(
+    if (capacity == 0) {
+      new LinkedBlockingQueue[InboxMessage]
+    } else {
+      new LinkedBlockingQueue[InboxMessage](capacity)
+    }
+  }
+
+  private val messageCount = new AtomicLong(0)
 
   // OnStart should be the first message to process
   inbox.synchronized {
     messages.add(OnStart)
   }
 
-  /**
-   * Process stored messages.
-   */
-  def process(dispatcher: Dispatcher): Unit = {
-    var message: InboxMessage = null
+  def stop(): Unit = inbox.synchronized {
+    // The following codes should be in `synchronized` so that we can make sure "OnStop" is the last
+    // message
+    if (!stopped) {
+      // We should disable concurrent here. Then when RpcEndpoint.onStop is called, it's the only
+      // thread that is processing messages. So `RpcEndpoint.onStop` can release its resources
+      // safely.
+      enableConcurrent = false
+      stopped = true
+      addMessage(OnStop)
+      // Note: The concurrent events in messages will be processed one by one.
+    }
+  }
+
+  def post(message: InboxMessage): Unit = {
+    if (stopped) {
+      // We already put "OnStop" into "messages", so we should drop further messages
+      onDrop(message)
+    } else {
+      addMessage(message)
+    }
+  }
+
+  // exposed only for testing
+  def getNumActiveThreads: Int = {
     inbox.synchronized {
-      if (!enableConcurrent && numActiveThreads != 0) {
-        return
-      }
-      message = messages.poll()
-      if (message != null) {
-        numActiveThreads += 1
-      } else {
-        return
-      }
+      inbox.numActiveThreads
     }
-    while (true) {
-      safelyCall(endpoint, endpointRef.name) {
-        message match {
-          case RpcMessage(_sender, content, context) =>
-            try {
-              endpoint.receiveAndReply(context).applyOrElse[Any, Unit](
-                content,
-                { msg =>
-                  throw new CelebornException(s"Unsupported message $message from ${_sender}")
-                })
-            } catch {
-              case e: Throwable =>
-                context.sendFailure(e)
-                // Throw the exception -- this exception will be caught by the safelyCall function.
-                // The endpoint's onError function will be called.
-                throw e
-            }
+  }
 
-          case OneWayMessage(_sender, content) =>
-            endpoint.receive.applyOrElse[Any, Unit](
-              content,
-              { msg =>
-                throw new CelebornException(s"Unsupported message $message from ${_sender}")
-              })
-
-          case OnStart =>
-            endpoint.onStart()
-            if (!endpoint.isInstanceOf[ThreadSafeRpcEndpoint]) {
-              inbox.synchronized {
-                if (!stopped) {
-                  enableConcurrent = true
-                }
-              }
-            }
+  def nextMessage(quitWithNoProcessingThread: Boolean): Option[InboxMessage] = {
```

Comment:
I think it's not that necessary to extract this method, and `quitWithNoProcessingThread` is a little confusing.